File: Year02/2nd semester/MPP/Lab/mpp-proiect-csharp-AlexandruPaul21/NetworkingProject/networking/ProjectClientWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;
using services;
using model;

namespace networking
{
    public class ProjectClientWorker : IProjectObserver //, Runnable
    {
        private IProjectServices server;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;
        public ProjectClientWorker(IProjectServices server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void notifyAgencies()
        {
            sendResponse(new BookingUpdate());
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;

            if (request is LoginRequest)
            {
                Console.WriteLine("Login request...");
                LoginRequest loginRequest = (LoginRequest)request;
                AgencyDto agencyDto = loginRequest.agencyDto;
                Agency agency = DtoUtils.fromDto(agencyDto);
                try
                {
                    lock (server)
                    {
                        server.login(agency, this);
                    }
                    return new OkResponse();
                }
                catch (ProjectException e)
                {
                    connected = false;
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is LogoutRequest)
            {
                Console.WriteLine("Logout request...");
                LogoutRequest loginRequest = (LogoutRequest)request;
                AgencyDto agencyDto = loginRequest.agencyDto;
                Agency agency = DtoUtils.fromDto(agencyDto);
                try
                {
                    lock (server)
                    {
                        server.logout(agency, this);
                    }
                    connected = false;
                    return new OkResponse();
                }
                catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is AllFlightsRequest)
            {
                Console.WriteLine("All flights request");
                AllFlightsRequest all = (AllFlightsRequest)request;
                AgencyDto agencyDto = all.agency;
                Agency agency = DtoUtils.fromDto(agencyDto);
                try
                {
                    Flight[] flights;
                    lock (server)
                    {
                        flights = server.getAllFlights(agency, this);
                    }

                    FlightDto[] flightDtos = DtoUtils.getDto(flights);

                    return new FlightsResponse(flightDtos);
                } catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is FilteredFlightsRequest)
            {
                Console.WriteLine("Filtered flights requested");
                FilteredFlightsRequest all = (FilteredFlightsRequest) request;

                FlightDto flightDto = all.flight;
                Flight flight = DtoUtils.fromDto(flightDto);
                try
                {
                    Flight[] flights;
                    lock (server)
                    {
                        flights = server.getFlightsForDestinationAndDate(flight, this);
                    }

                    FlightDto[] flightDtos = DtoUtils.getDto(flights);

                    return new FlightsResponse(flightDtos);
                }
                catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is NumberOfSeatsRequest)
            {
                Console.WriteLine("Number of seats request...");
                NumberOfSeatsRequest nrS = (NumberOfSeatsRequest)request;

                FlightDto flight = nrS.flight;
                Flight flight1 = DtoUtils.fromDto(flight);

                try
                {
                    int ans = 0;
                    lock (server)
                    {
                        ans = server.getNumberOfSeats(flight1, this);
                    }

                    return new SeatsResponse(ans);
                } 
                catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            if (request is BookingRequest)
            {
                Console.WriteLine("Booking request...");
                BookingRequest bookingRequest = (BookingRequest)request;

                Booking booking = DtoUtils.fromDto(bookingRequest.booking);

                try
                {
                    lock (server)
                    {
                        server.bookFlight(booking, this);
                    }
                    notifyAgencies();
                    return new OkResponse();
                }
                catch (ProjectException e)
                {
                    return new ErrorResponse(e.Message);
                }
            }

            return response;
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }

        }
    }


}

File: Year02/2nd semester/MPP/Lab/mpp-proiect-java-AlexandruPaul21/JavaProjectNetworking/ProjectNetworking/src/main/java/org/example/protocol/ProjectClientWorker.java
package org.example.protocol;

import org.example.*;
import org.example.dto.AgencyDto;
import org.example.dto.BookingDto;
import org.example.dto.DtoUtils;
import org.example.dto.FlightDto;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;

public class ProjectClientWorker implements Runnable, IProjectObserver {
    private IProjectServices server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ProjectClientWorker(IProjectServices server, Socket connection) {
        this.server = server;
        this.connection = connection;

        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while(connected){
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request)request);
                if (response != null){
                    sendResponse(response);
                }
            } catch (IOException | ClassNotFoundException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }

    private void sendResponse(Response response) throws IOException {
        output.writeObject(response);
        output.flush();
    }

    private static Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private Response handleRequest(Request request) {
        if (request.type() == RequestType.LOGIN) {
            AgencyDto dto = (AgencyDto)request.data();
            Agency agency = DtoUtils.fromDto(dto);
            try {
                server.login(agency, this);
                return okResponse;
            } catch (ProjectException e) {
                connected = false;
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.type() == RequestType.LOGOUT) {
            AgencyDto dto = (AgencyDto)request.data();
            Agency agency = DtoUtils.fromDto(dto);
            try {
                server.logout(agency, this);
                connected = false;
                return okResponse;
            } catch (ProjectException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.type() == RequestType.GET_ALL_FLIGHTS) {
            AgencyDto dto = (AgencyDto)request.data();
            Agency agency = DtoUtils.fromDto(dto);
            try {
                Flight[] ans = server.getAllFlights(agency, this);
                FlightDto[] dtos = DtoUtils.getDto(ans);
                return new Response.Builder().type(ResponseType.FLIGHTS).data(dtos).build();
            } catch (ProjectException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.type() == RequestType.GET_FILTERED_FLIGHTS) {
            FlightDto flightDto = (FlightDto)request.data();
            Flight flight = DtoUtils.fromDto(flightDto);

            try {
                Flight[] ans = server.getFlightsForDestinationAndDate(flight, this);
                FlightDto[] dtos = DtoUtils.getDto(ans);
                return new Response.Builder().type(ResponseType.FILTERED_FLIGHTS).data(dtos).build();
            } catch (ProjectException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.type() == RequestType.GET_NUMBER_OF_SEATS) {
            FlightDto flightDto = (FlightDto)request.data();
            Flight flight = DtoUtils.fromDto(flightDto);

            try {
                int ans = server.getNumberOfSeats(flight, this);
                return new Response.Builder().type(ResponseType.NUMBER_SEATS).data(ans).build();
            } catch (ProjectException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        if (request.type() == RequestType.BOOK_TICKET) {
            BookingDto bookingDto = (BookingDto)request.data();
            Booking booking = DtoUtils.fromDto(bookingDto);

            try {
                server.bookFlight(booking, this);
                return okResponse;
            } catch (ProjectException e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }

        return null;
    }

    @Override
    public void ticketBuyed() throws ProjectException {
        AgencyDto agencyDto = new AgencyDto("agency1", "agency1","1234");
        Response response = new Response.Builder().type(ResponseType.BOOKING).data(agencyDto).build();

        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new ProjectException("Error sending response "+e);
        }
    }
}
